### PR TITLE
Adding logsoftmax

### PR DESCRIPTION
**Pull request by boathit**
Adding `logsoftmax` function by following the convention of `softmax`.

### CODE REVIEW

**Comment by MikeInnes** [comment]
Awesome, thanks! Can you add a test for this? You can perhaps compare it to regular `softmax` to make sure it's correct.

**Comment by boathit** [comment]
Sure. Actually, I have compared it with the PyTorch implementation `torch.nn.functional.log_softmax()`  and the outputs are consistent. I will add several test cases in `test.jl`.

**Comment by MikeInnes** [comment]
Great, thanks!

Just FYI, this won't be usable with Flux until you add gradient definitions there too; happy to give pointers there if you need them.

**Comment by boathit** [comment]
Hi @MikeInnes, I indeed added the gradient definition, see the line `∇logsoftmax(Δ, xs) = ∇softmax(Δ ./ softmax(xs), xs)`

**Comment by MikeInnes** [comment]
Sure, that's great; but Flux won't know about the gradient unless you tell it, so you need a definition [something like this](https://github.com/FluxML/Flux.jl/blob/b0b8c9dbd1303e6363a881553175bb7d1ebdceb7/src/tracker/lib.jl#L137) on the Flux side.

**Comment by boathit** [comment]
Oh, I see. Perhaps I can make another PR for Flux. Besides, even though the current `logsoftmax()` works with CuArrays but I am not sure whether CUDNN has more efficient GPU version.

**Comment by MikeInnes** [comment]
Yeah, this will be very slow for CuArrays. For softmax we [add CUDNN overloads](https://github.com/JuliaGPU/CuArrays.jl/blob/00276e3c234188d728a94c12d88674839de642dd/src/dnn/nnlib.jl#L10-L18) in CuArrays itself, and the same thing could be done for `logsoftmax`.

**Comment by boathit** [comment]
Hi @MikeInnes, as spoken a [PR #153](https://github.com/FluxML/Flux.jl/pull/153) has been made to Flux.jl to register backward function for `logsoftmax`.

**Comment by MikeInnes** [comment]
:+1: Thanks!

File: src/logsoftmax.jl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      xi_max = xs[1, j]
+      for i = 1:size(out, 1)
+        xi_max = max(xi_max, xs[i, j])
+      end
+      s = zero(eltype(out))
+      for i = 1:size(out, 1)
+        s += exp(xs[i, j] - xi_max)
+      end
+      for i = 1:size(out, 1)
+        out[i, j] = xs[i, j] - log(s) - xi_max
```
Comment (vchuravy) [inline src/logsoftmax.jl]: Do you remember why you subtract `xi_max` here?
IIRC the definition of logsoftmax normally is

```
x_max = maximum(xs)
s = log(sum(exp.(xs - x_max)))
out = xs .- s
```


```

Comment (boathit) [reply src/logsoftmax.jl]: I think we are doing the same thing as pytorch. If you replace `s` with `log(sum(exp.(xs - x_max))) + x_max` you will get 
```
out = xs .- log(sum(exp.(xs - x_max))) - x_max
```
